File: protocol/src/main/resources/java/ProtocolManagerTemplate.java
${protocol_root_path}
import java.util.HashMap;
import java.util.Map;
${protocol_imports}

public class ProtocolManager {

    public static final short MAX_PROTOCOL_NUM = Short.MAX_VALUE;

    public static final IProtocolRegistration[] protocols = new IProtocolRegistration[MAX_PROTOCOL_NUM];

    public static Map<Class<?>, Short> protocolIdMap = new HashMap<>();

    public static void initProtocol() {
        // initProtocol
        ${protocol_manager_registrations}
    }

    public static short getProtocolId(Class<?> clazz) {
        return protocolIdMap.get(clazz);
    }

    public static IProtocolRegistration getProtocol(short protocolId) {
        var protocol = protocols[protocolId];
        if (protocol == null) {
            throw new RuntimeException("[protocolId:" + protocolId + "] not exist");
        }
        return protocol;
    }

    public static void write(ByteBuffer buffer, Object packet) {
        var protocolId = getProtocolId(packet.getClass());
        // write protocol id to buffer
        buffer.writeShort(protocolId);
        // write packet
        getProtocol(protocolId).write(buffer, packet);
    }

    public static Object read(ByteBuffer buffer) {
        var protocolId = buffer.readShort();
        return getProtocol(protocolId).read(buffer);
    }

}

File: protocol/src/main/resources/kotlin/ProtocolManagerTemplate.kt
${protocol_root_path}
${protocol_imports}
class ProtocolManager {
    companion object {
        val protocols = arrayOfNulls<IProtocolRegistration>(Short.MAX_VALUE.toInt())
        var protocolIdMap: MutableMap<Class<*>, Short> = HashMap()
        @JvmStatic
        fun initProtocol() {
            // initProtocol
            ${protocol_manager_registrations}
        }

        @JvmStatic
        fun getProtocolId(clazz: Class<*>): Short {
            return protocolIdMap[clazz]!!
        }

        @JvmStatic
        fun getProtocol(protocolId: Short): IProtocolRegistration {
            return protocols[protocolId.toInt()]
                ?: throw RuntimeException("[protocolId:$protocolId] not exist")
        }

        @JvmStatic
        fun write(buffer: ByteBuffer, packet: Any) {
            val protocolId = getProtocolId(packet.javaClass)
            // write protocol id to buffer
            buffer.writeShort(protocolId)
            // write packet
            getProtocol(protocolId).write(buffer, packet)
        }

        @JvmStatic
        fun read(buffer: ByteBuffer): Any {
            val protocolId = buffer.readShort()
            return getProtocol(protocolId).read(buffer)
        }
    }
}